Hilbert series

HechtiDerLachs opened the pull request:
This adds the (fast) implementation of different strategies for the computation of multivariate Hilbert series.

HechtiDerLachs committed:
First implementation of a different strategy for multivariate Hilbert series.

HechtiDerLachs committed:
Implement all the strategies from the book.

HechtiDerLachs committed:
Some fixes.

HechtiDerLachs committed:
WIP on Hilbert polynomials.

HechtiDerLachs committed:
WIP.

HechtiDerLachs committed:
WIP on Bayer-Stillman

HechtiDerLachs committed:
Finish Bayer-Stillman.

HechtiDerLachs committed:
Add some tests.

HechtiDerLachs committed:
Add binding of the backup to the user-facing method and documentation.

HechtiDerLachs committed:
Add missing semi colon.

tthsqe12 commented:
Instead of `strategy`, we have been using `alg` as the name of the keywork argument for selecting algorithms

fingolfin commented:
I agree that for now we should match what is there and use `alg`. But I also think that we should replace `alg` by `algorithm` (or `strategy` or whatever) following our general naming rules... I'll open an issue for that

HechtiDerLachs committed:
strategy -> alg

fingolfin commented on docs/src/CommutativeAlgebra/affine_algebras.md:
```diff
@@ -484,7 +484,7 @@ degree(A::MPolyQuo)
-multi_hilbert_function(A::MPolyQuo, g::GrpAbFinGenElem)
+  multi_hilbert_series(A::MPolyQuo; strategy::Symbol=:BayerStillmanA)
+  multi_hilbert_series_reduced(A::MPolyQuo; strategy::Symbol=:BayerStillmanA)
+  multi_hilbert_function(A::MPolyQuo, g::GrpAbFinGenElem)
```
```suggestion
multi_hilbert_series(A::MPolyQuo; alg::Symbol=:BayerStillmanA)
multi_hilbert_series_reduced(A::MPolyQuo; alg::Symbol=:BayerStillmanA)
multi_hilbert_function(A::MPolyQuo, g::GrpAbFinGenElem)
```

HechtiDerLachs replied:
Oops... Did I forget them? Thanks for catching this!

fingolfin reviewed (commented):
One minor nitpick. Otherwise I think this waiting for a review by @hannes14 resp. suggestions how to do the computations more efficiently, to catch up with Singular's performance

HechtiDerLachs committed:
Update docs/src/CommutativeAlgebra/affine_algebras.md

Co-authored-by: Max Horn <max@quendi.de>

hannes14 reviewed (approved):
Most of the time is spent in _divide_by: Singular uses "short exponent vector": a 64bit word,
where bit i is set if exponent i >0. If short_exp_vector(a) & not short_exp_vector(b)<>0 (bitwise)
then monomial a cannot divide monomial b.
This pre-test results in a speed-up factor of 5 in Singular

fingolfin commented:
@hannes14 thanks for the info. I guess we could implement such a test in Julia, too, at least I see no fundamental reason against it?

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,349 @@ function Base.show(io::IO, h::HilbertData)
+  # the good ones.
+  bad = [e for e in shifted if !all(x->(x>=0), e)]
+
+  divides(a::Vector{Int}, b::Vector{Int}) = all(k->(k>=0), b - a)
```
this computes (and thus allocates) the vector `b-a`, then runs over it to check the entries for being positive. I am guessing this will be quite a bit slower than doing the direct approach, i.e.,
```suggestion
  divides(a::Vector{Int}, b::Vector{Int}) = all(k->b[k]>=a[k], 1:length(a))
```
Indeed:
```
julia> v1 = collect(1:10) ; v2 = collect(1:10) * 2;

julia> @btime divides(v1,v2) ; @btime divides2(v1,v2);
  30.402 ns (1 allocation: 144 bytes)
  8.133 ns (0 allocations: 0 bytes)

julia> @btime divides(v2,v1) ; @btime divides2(v2,v1);
  26.800 ns (1 allocation: 144 bytes)
  4.333 ns (0 allocations: 0 bytes)
```

thofma replied:
Or `all((>=).(b, a))` for maximal julia style and compactness (but it is a bit slower, since unfortunately there is no "lazy" version).

fingolfin replied:
That's really elegant and compact, but it performs worst of all the three options in my test (in Julia 1.8.1 on an M1 mac):
```
julia> using BenchmarkTools

julia> divides(a::Vector{Int}, b::Vector{Int}) = all(k->(k>=0), b - a);

julia> divides2(a::Vector{Int}, b::Vector{Int}) = all(k->b[k]>=a[k], 1:length(a));

julia> divides3(a::Vector{Int}, b::Vector{Int}) = all((>=).(b, a));

julia> divides4(a::Vector{Int}, b::Vector{Int}) = @inbounds all(k->b[k]>=a[k], 1:length(a));

julia> v1 = collect(1:10) ; v2 = collect(1:10) * 2;

julia> @btime divides(v1,v2) ; @btime divides2(v1,v2); @btime divides3(v1,v2);
  30.402 ns (1 allocation: 144 bytes)
  8.133 ns (0 allocations: 0 bytes)
  40.657 ns (2 allocations: 96 bytes)

julia> @btime divides(v2,v1) ; @btime divides2(v2,v1); @btime divides3(v2,v1);
  26.842 ns (1 allocation: 144 bytes)
  4.375 ns (0 allocations: 0 bytes)
  40.616 ns (2 allocations: 96 bytes)

julia> @btime divides(v1,v1) ; @btime divides2(v1,v1); @btime divides3(v1,v1);
  30.977 ns (1 allocation: 144 bytes)
  3.083 ns (0 allocations: 0 bytes)
  40.573 ns (2 allocations: 96 bytes)
```

BTW these numbers are from a "pure" Julia session, no Oscar loaded. For reasons I can't fathom, variants 1 and 3 are quite a bit slower with Oscar loaded, even though the machine code according to `@code_native` appears to be identical. Huh.

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,349 @@ function Base.show(io::IO, h::HilbertData)
+  # generator which is collected in `bad` a priori. It is checked 
+  # whether they become superfluous and if not, they are added to 
+  # the good ones.
+  bad = [e for e in shifted if !all(x->(x>=0), e)]
```
to compute good and bad you go over the same array twice, computing the exact same things, simply discarding other items. I expect that you can gain close to a factor two in performance by merging this into a single loop.

Indeed, consider this input:
```julia
function foo(shifted)
    good = [e for e in shifted if all(x->(x>=0), e)]
    bad = [e for e in shifted if !all(x->(x>=0), e)]
    return good, bad
end

function foo2(shifted)
    good = similar(shifted, 0)
    bad = similar(shifted, 0)
    for e in shifted
        if all(x->(x>=0), e)
            push!(good, e)
        else
            push!(bad, e)
        end
    end
    return good, bad
end

a = [ 1:10,  (1:10) * -1,  (1:10) * 2,  (1:10) - ones(Int, 10), zeros(Int, 10), ones(Int, 10) ]
pivot = ones(Int, 10)
shifted = [collect(e-pivot) for e in a]
```
Then we observe this:
```
julia> foo(shifted) == foo2(shifted)
true

julia> @btime foo(shifted); @btime foo2(shifted);
  275.272 ns (25 allocations: 960 bytes)
  127.552 ns (5 allocations: 288 bytes)
```
So it is actually more than a factor 2, and the allocations have reduced substantially as well. One can probably improve it further

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,349 @@ function Base.show(io::IO, h::HilbertData)
+  for e in bad 
+    m = [k < 0 ? 0 : k for k in e]
+    if all(x->(!divides(x, m)), good)
+      good = [x for x in good if !divides(m, x)]
```
This is a point where keeping those Int64 bit masks as a pre-divisibility test, as described by @hannes14, would probably have a major effect.

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,349 @@ function Base.show(io::IO, h::HilbertData)
+########################################################################
+exp_vec(f::MPolyElem) = first(exponent_vectors(f))
+
+_divides(a::Vector{Int}, b::Vector{Int}) = all(x->(x>=0), a-b)
```
this, too, of course should be optimized as sketched elsewhere (also, if you have this, why define it again? That just adds compiler over head?)

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,349 @@ function Base.show(io::IO, h::HilbertData)
+# [Kreuzer, Robbiano: Computational Commutative Algebra 2, Springer]
+# Section 5.3.
+########################################################################
+exp_vec(f::MPolyElem) = first(exponent_vectors(f))
```
Isn't this the same as `leading_exponent_vector` ?

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,349 @@ function Base.show(io::IO, h::HilbertData)
+
+_divides(a::Vector{Int}, b::Vector{Int}) = all(x->(x>=0), a-b)
+
+function _gcd(a::Vector{Vector{Int}}) 
```
You only ever use this for vectors `a` of length two, so perhaps write a function that takes two `Vector{Int}` instead? that could then potentially be quite a bit quicker.

HechtiDerLachs replied:
The new method for two arguments is now implemented below.

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,349 @@ function Base.show(io::IO, h::HilbertData)
+  t = gens(return_ring)
+
+  # See Proposition 5.3.6
+  _are_pairwise_coprime(a) && return prod([1-prod([t[i]^(sum([e[j]*weight_matrix[i, j] for j in 1:length(e)])) for i in 1:length(t)]) for e in a])
```
I would advise against writing such long convoluted lines. They are hard to read, understand and debug. As shown in other comments, often a simple loop is also faster; that may not be the case here, but I would recommend to at least change it to an `if ... end` and then apply some line wrapping

HechtiDerLachs replied:
This has been done at some internal point below. Since we're waiting for the MPolyBuildCtx, this is subject to change, anyway and can be copied from there when the time is ripe.

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,349 @@ function Base.show(io::IO, h::HilbertData)
+                                                    weight_matrix=weight_matrix, 
+                                                    alg=alg)
+      for k in linear_mons
+        q = q*(one(S) - prod([t[i]^weight_matrix[i, k] for i in 1:length(t)]))
```
The following may or may not matter here (measure it using `@btime` ?), but: instead of writing `prod([x for y in z])` you can also write `prod(x for y in z)`. The latter avoids allocating a vector and storing all intermediate results. Whether this is faster or not depends on various factors, but it can be quite substantially if the length (here: `length(t)`) is big.

HechtiDerLachs committed:
WIP.

HechtiDerLachs commented:
Thanks for the input! I will have a look this afternoon.

HechtiDerLachs committed:
First changes from discussion.

HechtiDerLachs committed:
Implement suggestions from discussion.

HechtiDerLachs commented:
I get no speedup from either of the suggestions above. Still singular is about ten times faster. Probably I am still doing something wrong, but I did my best for the time being. Only that this did not really increase the readability of the code...

I would like to point out, however, that it was my decision to perform all these operations on intvec's rather than on monomials given as `MPolyElem`s. So I guess there is still a significant speedup due to my work.

thofma commented:
Can you post the julia code for the example that you tried for the timings?

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,386 @@ function Base.show(io::IO, h::HilbertData)
+    # affirmative case and pivot can not divide e.
+    if iszero(short_piv & ~ s) && all(k->(e[k]>=pivot[k]), 1:length(e)) # tuned version of divides(e, pivot)
+      push!(good, next)
+      push!(good_shorts, sum([(next[k]>0 ? 1 : 0) << (k-1) for k in 1:length(next)]))
```
Did you try first applying my hints then measure, before adding these "shorts"? Because that code is again extremely inefficient and in this form likely constitutes a slowdown, not a speed up...


(The whole idea was to use bitmasks in an Int to be super fast; adding more vectors like this is several orders of magnitude slower. I would suggest to remoece the "shorts" again completely for now, and later Tommy or me can perhaps help adding them back in the intended form)

fingolfin replied:
To explain what is bad about this, observe that *must* allocate a `Vector{Int}`, which then is summed up. This takes up a ton of memory. Optimal code would "allocate" a single `Int` and then directly set bits in it.

Alas, let's not get hold up on this, let's just remove it for now. I'll write more in a longer comment in a moment

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,386 @@ function Base.show(io::IO, h::HilbertData)
+    end
+  end
+
+  for (e, s) in zip(bad, bad_shorts)
```
Don't use zip

HechtiDerLachs commented:
Test code:
```julia
g = [[23, 43, 25, 24, 20], [40, 27, 20, 46, 20], [22, 20, 31, 24, 39], [48, 35, 21, 21, 22], [22, 23, 27, 38, 20], [23, 28, 20, 34, 24], [21, 20, 26, 44, 36], [24, 34, 21, 33, 22], [21, 21, 23, 50, 27], [43, 30, 30, 20, 21], [20, 39, 22, 27, 27], [24, 25, 22, 37, 38], [20, 21, 31, 31, 20], [21, 20, 26, 41, 46], [22, 36, 24, 33, 22], [25, 24, 21, 26, 22], [20, 20, 30, 42, 37], [20, 22, 37, 30, 33], [26, 31, 27, 20, 21], [46, 22, 24, 20, 32], [23, 24, 26, 20, 48], [23, 44, 45, 20, 20], [20, 50, 23, 42, 21], [29, 21, 20, 21, 25], [23, 29, 21, 32, 36], [23, 21, 21, 37, 42], [44, 20, 20, 50, 25], [30, 21, 21, 20, 40], [20, 36, 26, 25, 28], [21, 22, 20, 39, 36], [21, 24, 31, 24, 41], [20, 38, 24, 33, 25], [23, 39, 21, 28, 22], [30, 48, 20, 20, 31], [25, 34, 23, 20, 45], [21, 26, 21, 42, 25], [23, 35, 21, 32, 21], [22, 20, 33, 22, 45], [21, 34, 29, 28, 35], [22, 24, 26, 21, 49], [22, 22, 44, 20, 32], [20, 49, 21, 31, 24], [20, 39, 28, 21, 26], [22, 21, 24, 23, 45], [23, 46, 24, 21, 22], [21, 35, 27, 41, 23], [45, 39, 20, 23, 21], [22, 36, 29, 20, 39], [50, 25, 21, 28, 20], [20, 29, 39, 20, 44], [25, 22, 24, 40, 20], [22, 22, 27, 21, 21], [38, 21, 20, 30, 22], [20, 38, 30, 45, 21], [24, 20, 24, 32, 29], [21, 36, 31, 23, 30], [21, 26, 31, 23, 34], [46, 23, 20, 25, 24], [21, 29, 31, 24, 28], [21, 43, 23, 29, 20], [49, 23, 42, 20, 21], [21, 41, 25, 20, 25], [21, 49, 22, 29, 24], [28, 27, 43, 20, 25], [29, 22, 28, 20, 33], [21, 20, 35, 21, 27], [31, 46, 25, 20, 20], [37, 34, 21, 24, 20], [22, 39, 23, 27, 20], [31, 41, 20, 50, 23], [20, 40, 22, 23, 46], [45, 22, 24, 22, 20], [22, 39, 22, 46, 20], [24, 20, 23, 38, 26], [20, 40, 26, 20, 46], [27, 20, 38, 20, 39], [37, 21, 48, 20, 23], [25, 35, 45, 20, 24], [44, 35, 20, 20, 42], [26, 20, 22, 31, 29], [21, 44, 27, 20, 24], [27, 22, 22, 27, 20], [25, 46, 22, 34, 20], [22, 47, 23, 20, 20], [20, 20, 36, 28, 47], [23, 26, 21, 34, 35], [21, 34, 28, 20, 48], [26, 37, 45, 23, 20], [29, 21, 22, 50, 20], [20, 21, 34, 23, 36], [50, 21, 30, 20, 23], [32, 22, 24, 22, 22], [25, 20, 29, 29, 20], [36, 20, 23, 21, 48], [24, 33, 28, 20, 45], [20, 29, 21, 35, 40], [46, 22, 36, 20, 22], [44, 34, 23, 21, 23], [34, 42, 23, 21, 22], [39, 22, 20, 29, 23], [20, 24, 48, 20, 20], [48, 26, 23, 20, 39], [49, 22, 35, 20, 23], [39, 21, 49, 23, 22], [23, 40, 46, 20, 21], [29, 20, 23, 24, 33], [36, 20, 31, 21, 29], [20, 20, 33, 47, 35], [33, 20, 21, 34, 29], [23, 24, 23, 25, 20], [21, 40, 23, 27, 22], [20, 31, 34, 21, 21], [45, 20, 21, 22, 22], [42, 28, 22, 20, 36], [22, 25, 22, 31, 48], [21, 22, 40, 24, 25], [42, 26, 21, 22, 23], [28, 31, 20, 20, 45], [20, 24, 23, 47, 25], [25, 27, 20, 26, 39], [28, 22, 28, 20, 45], [22, 20, 23, 38, 46], [20, 23, 34, 24, 22], [24, 32, 20, 23, 24], [46, 21, 33, 25, 20], [45, 33, 21, 45, 21], [42, 20, 47, 22, 26], [21, 24, 30, 28, 42], [27, 23, 28, 21, 20], [25, 20, 20, 45, 35], [20, 38, 27, 21, 50], [24, 22, 23, 46, 25], [22, 47, 22, 20, 21], [21, 36, 20, 28, 38], [26, 25, 20, 39, 32], [26, 21, 42, 24, 24], [20, 27, 32, 22, 22], [22, 23, 23, 22, 24], [23, 21, 31, 20, 27], [21, 35, 25, 20, 43], [20, 20, 41, 28, 32], [21, 21, 20, 50, 48], [32, 36, 32, 20, 20], [20, 23, 28, 33, 30], [34, 40, 20, 20, 37], [20, 27, 23, 22, 39], [43, 20, 24, 33, 28], [20, 20, 33, 23, 50], [27, 21, 20, 26, 42], [23, 23, 21, 46, 21], [44, 20, 20, 25, 30], [20, 47, 23, 27, 23], [31, 27, 35, 20, 26], [30, 30, 30, 20, 30], [20, 20, 35, 43, 20], [43, 25, 20, 48, 21], [33, 20, 26, 20, 44], [22, 40, 39, 20, 37], [22, 20, 43, 26, 22], [20, 29, 20, 36, 35], [21, 26, 20, 36, 36], [21, 42, 30, 23, 21], [35, 38, 21, 20, 26], [37, 38, 20, 28, 20], [20, 24, 24, 45, 48], [21, 27, 22, 30, 25], [27, 21, 21, 37, 23], [39, 20, 27, 31, 27], [38, 34, 24, 22, 20], [35, 39, 20, 20, 24], [22, 31, 21, 48, 23], [26, 20, 22, 48, 28], [20, 22, 30, 46, 31], [21, 20, 27, 41, 41], [21, 25, 24, 29, 47], [20, 26, 45, 23, 34], [22, 23, 22, 43, 29], [42, 22, 30, 24, 20], [30, 45, 20, 22, 20], [20, 31, 20, 31, 28], [20, 26, 22, 37, 26], [20, 23, 41, 22, 37]]

# singular timings
  R, x = PolynomialRing(QQ, ["x$i" for i in 1:5])
  P, _ = grade(R)
  I = ideal(P, [prod([x[i]^e[i] for i in 1:length(x)]) for e in g])
  Q, _ = quo(P, I)

  @time hilbert_series(Q)

# Oscar timings
 W = [1 1 1 1 1]
@time Oscar._hilbert_numerator_from_leading_exponents(g, weight_matrix=W, alg=:BayerStillmanA)
```
No, I did not write down the measurements before implementing the `shorts`. I just remembered that it used to be a factor 10 in total and it still is a factor ten. If it had been significantly different with the intermediate step, I would have noticed. In particular, there was no double speedup because of removing the repeated loop. 

I don't see how my code is 'extremely inefficient'. I mean: The bitmask vectors have to be dragged along somehow and I decided for an extra array in order to keep the compatibility with the previous version. If I put the bitmasks together with the exponent vectors, would that make a difference? Because I could avoid using `zip`? 

For most of the previous suggestions, I could have come up with them myself or at least understand why they are good tricks. But now this seems to go beyond my expertise since I have no good idea why stuff like `zip` is a performance pitfall. Because it allocates a tuple for each pair of entries? I would expect the julia compiler to recognize a workaround for that. 

However, I am curious to see how this must be done to really make a difference. So let me know if you made it.

ederc commented:
As far as I can see your choice for the short divisor mask bits is not the best for your given example: If you set the bits for an exponent vector entry if the exponent is `>0` then in most of your `_divide_by` tests `iszero(short_piv & ~ s)` will return `true`, even so `pivot` is not a divisor of `e`. In other words: Your effective hit rate is too poor. Running your example I get in nearly `90%` of the divisibility tests a `true` from your short divisor check and a `false` then from the full check on the exponents.

I would suggest that you decide the setting of the bits in the short divisor masks via a different measure. For example, in `msolve` I calculate the mean of maximal and minimal exponent for a given variable in the leading monomials at a given stage of the computation (e.g. min and max exponents for each variable for all leading monomials of the input generators) and set the bits in the short divisor mask if the exponent of the exponent in the given monomial is greater or equal to this mean.

On the other hand you will always have the overhead of precomputing the short divisor masks in your specific code, in `Singular`, `msolve`, `magma`, etc. those short divisor masks are stored directly for each handled monomial, i.e. they are somewhat part of the monomial data structure.

thofma commented on test/Rings/Hilbert_series_test.jl:
```diff
@@ -0,0 +1,19 @@
+# To be put in a proper test suite.
+R, (x,y,z,w) = QQ["x", "y", "z", "w"]
+I = ideal(R, [z^3, y^3*z*w^2, x^2*y^4*w^2, x*y*z^2])
+@show Oscar._hilbert_numerator_from_leading_ideal(I)
```
```suggestion
Oscar._hilbert_numerator_from_leading_ideal(I)
```

thofma commented on test/Rings/Hilbert_series_test.jl:
```diff
@@ -0,0 +1,19 @@
+
+I = ideal(P, [X^2, Y, Z^3])
+
+@show Oscar._hilbert_numerator_from_leading_ideal(I)
```
```suggestion
Oscar._hilbert_numerator_from_leading_ideal(I)
```

fingolfin committed:
Remove 'shorts' code, optimizations

fingolfin committed:
Use filter!

fingolfin committed:
Avoid another allocation

fingolfin commented:
Measuring with `@time` is not that useful, at least if only done once, as it includes compile time. Better to use `@btime` from `BenchmarkTools`. I wrapped you examples into helper functions, to (hopefully) avoid caching
```julia
function hilb_with_sing(g)
  # singular timings
  R, x = PolynomialRing(QQ, ["x$i" for i in 1:5])
  P, _ = grade(R)
  I = ideal(P, [prod([x[i]^e[i] for i in 1:length(x)]) for e in g])

  Q, _ = quo(P, I)
  return hilbert_series(Q)
end

function hilb_with_oscar(g)
  W = [1 1 1 1 1]
  Oscar._hilbert_numerator_from_leading_exponents(g, weight_matrix=W, alg=:BayerStillmanA)
end
```
Then I get these timings on my M1 MacBook Pro:
```
julia> @btime hilb_with_sing(g);
  4.519 ms (32875 allocations: 1.11 MiB)

julia> @btime hilb_with_oscar(g);  # with previous HEAD of this PR, 8c21670ced5e7552a239506ab13622868161b844
  29.617 ms (899648 allocations: 54.87 MiB)

julia> @btime hilb_with_oscar(g);  # with current HEAD of this PR, 7009820932066cef8f8ea94d87f9ef08eb4dbfd4 
  23.925 ms (589138 allocations: 40.36 MiB)
```
So it went from being slower by a factor of 6.5 to "just" a factor 5.3. I then proceeded to remove the `shorts` code, upon which it went to this; note in particular the substantially reduced number of allocations and memory usage
```
julia> @btime hilb_with_oscar(g);
  18.286 ms (402734 allocations: 29.27 MiB)
```
There are more micro optimizations; e.g. moving the line `new_good = similar(good, 0)` a few lines down, into the `if`, so that we only allocate if we really need to, drops it further:
```
  18.105 ms (387834 allocations: 28.59 MiB)
```
I am only mentioning these intermediate steps to illustrate how seemingly tiny changes in such "hot" code can have serious consequences. As it is, we can do even better, by using `filter!` to completely avoid this allocation.
```
  17.170 ms (356948 allocations: 26.43 MiB)
```
With some more tweaking my (for now) final version is "only" slower by a factor of 3.77
```
  16.987 ms (348079 allocations: 25.61 MiB)
```
I've pushed that now. I see more things that could be optimized but I need sleep now. Also, so far I only looked at this one function... time to start the profiler again now, perhaps?

HechtiDerLachs commented:
Alright, thanks for the detailed insight. Especially the things like `filter!` were still missing for me. I had been wondering whether julia possesses anything like linked lists so that one would not be forced to reallocate stuff in every iteration. But I guess that methods like `filter!` make use of such things in the background. 

We could modify the input list by replacing every exponent intvec `e = [e_1, ..., e_n]` by an extended intvec `[b, e_1, ..., e_n]` where `b` is the bitmask for the original `e`. Then we could also make that part of the data structure. Otherwise: Can anybody estimate whether going back to `MPoly`s instead of intvecs would even improve the situation? Do they also make use of native bitmasking like singular does?

HechtiDerLachs commented:
@thofma This file should go away in total. It's a mistake that it was committed in the first place. The real test suite is in the tests for the graded rings.

fingolfin commented:
Don't worry about bitmasks or so at this point.

I somewhat doubt that using mpoly will help here; it was (and probably are) the allocations that are killing you, and those don't go away by switching between `Vector{Int}` and an `mpoly` (also, if this is just about representing a single exponent vector, I don't think an `mpoly`, which is much more general, could beat a plain vector; but then I didn't look or try, and don't know the mpoly code well, so I might be wrong -- in these things, the one thing I do know is that you should always measure measure measure. Instincts and gut feelings and things one "knows" can guide one in what to try, but they can't replace measurements.

Also, I worry this example may be too small for serious tuning, at just a few milliseconds runtime. It may well be that in big examples the bottlenecks are elsewhere.

HechtiDerLachs committed:
Remove default arguments from internal function.

HechtiDerLachs committed:
Take the specialized algorithms to their own functions.

HechtiDerLachs committed:
Repair type stability.

HechtiDerLachs committed:
Tune products and coprime-check.

HechtiDerLachs committed:
Add code for MPolyBuildCtx.

fingolfin commented:
We discussed various possible optimizations and more changes will likely follow here

HechtiDerLachs committed:
Add comment on status quo and possibilities for tuning in the future.

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,467 @@ function Base.show(io::IO, h::HilbertData)
+  n = length(a[1]) # assumed to be the length of all entries of a
+  return [minimum([a[i][j] for i in 1:length(a)]) for j in 1:n]
+end
+
```
```suggestion
```

HechtiDerLachs committed:
Remove exp_vec.

HechtiDerLachs committed:
Remove faulty test file.

fingolfin committed:
Remove unused _gcd variant

fingolfin committed:
Speed up _are_pairwise_coprime

Before:

    julia> v=[[0,2,0],[0,0,2],[2,0,0]] ; @btime Oscar._are_pairwise_coprime(v)
      90.925 ns (3 allocations: 240 bytes)
    true

After:

    julia> v=[[0,2,0],[0,0,2],[2,0,0]] ; @btime Oscar._are_pairwise_coprime(v)
      17.911 ns (0 allocations: 0 bytes)
    true

fingolfin committed:
Factor out some common code

fingolfin commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1612,6 +1612,467 @@ function Base.show(io::IO, h::HilbertData)
+    end
+  end
+  return vcat(less, [pivot], more)
+end
```
The one place calling this has a comment saying "make sure we have lexicographically ordered monomials"

But that does not match what this function does; it does *not* sort the vector by lex order (unless it has at most 2 elements, or by pure luck). Instead it does something weird.

I am going to assume the comment is correct, not the code. Then one can simply use `sort!` instead. Doing that doubles the speed of the code for me in a tiny example (BTW if you could tell me how to produce a bigger example, that would be helpful)

HechtiDerLachs replied:
True. Now when I look at it again, it does not make sense, at all. I remember, I wanted to implement quicksort for that. But probably, I got distracted half-way through and never finished in any reasonable way. Thanks again for spotting that! It is weird, though that the outputs of all random examples had always been correct, either way...

Edit: To produce bigger examples, just turn up the parameters in the test code [here](https://github.com/oscar-system/Oscar.jl/pull/1543/files#diff-b05943836d126d03a8db0fa0c59df7b0256fac98ce9ad11741a249ce02771701R255-R262).

fingolfin committed:
Really sort monomials lexicographically

Before:

    julia> @btime res == hilb_with_oscar(g);
      8.905 ms (135721 allocations: 11.29 MiB)

After:

    julia> @btime res == hilb_with_oscar(g);
      7.611 ms (110184 allocations: 9.54 MiB)

HechtiDerLachs commented on src/Rings/mpoly-graded.jl:
```diff
@@ -1718,28 +1718,6 @@ function _find_maximum(a::Vector{Int})
-  return true
-end
-
-function _sort_lex(a::Vector{Vector{Int}})
```
@fingolfin : The name of your commit suggests that there was an error in my code. Was it? 

Either way: This should be much better now. I didn't know `sort!` already does this. Thanks for cleaning it up.

fingolfin replied:
Sorry, I had actually made a comment with an explanation here but forgot to submit it. Will do that now

HechtiDerLachs committed:
Fix tests from local test file.

HechtiDerLachs committed:
Merge branch 'hilbert_series' of github.com:HechtiDerLachs/Oscar.jl into hilbert_series
